src/ifcgeom/mapping/IfcSegmentedReferenceCurve.cpp

```diff
@@ -0,0 +1,84 @@
+ * This file is part of IfcOpenShell.                                           *
+ *                                                                              *
+ * IfcOpenShell is free software: you can redistribute it and/or modify         *
+ * it under the terms of the Lesser GNU General Public License as published by  *
+ * the Free Software Foundation, either version 3.0 of the License, or          *
+ * (at your option) any later version.                                          *
+ *                                                                              *
+ * IfcOpenShell is distributed in the hope that it will be useful,              *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 *
+ * Lesser GNU General Public License for more details.                          *
+ *                                                                              *
+ * You should have received a copy of the Lesser GNU General Public License     *
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.         *
+ *                                                                              *
+ ********************************************************************************/
+
+#include "mapping.h"
+#define mapping POSTFIX_SCHEMA(mapping)
+using namespace ifcopenshell::geometry;
+
+#ifdef SCHEMA_HAS_IfcSegmentedReferenceCurve
+
+taxonomy::ptr mapping::map_impl(const IfcSchema::IfcSegmentedReferenceCurve* inst) {
+   if (!inst->BaseCurve()->as<IfcSchema::IfcGradientCurve>())
+       Logger::Warning("Expected IfcSegmentedReferenceCurve.BaseCurve to be IfcGradient", inst); // CT 4.1.7.1.1.3
+		  
+	auto gradient = taxonomy::cast<taxonomy::piecewise_function>(map(inst->BaseCurve()));
+	auto cant = taxonomy::make<taxonomy::piecewise_function>();
+
+	auto segments = inst->Segments();
+
+	for (auto& segment : *segments) {
+		if (segment->as<IfcSchema::IfcCurveSegment>()) {
+			// @todo check that we don't get a mixture of implicit and explicit definitions
+			auto crv = map(segment->as<IfcSchema::IfcCurveSegment>());
+			if (crv->kind() == taxonomy::PIECEWISE_FUNCTION) {
+				auto seg = taxonomy::cast<taxonomy::piecewise_function>(crv);
+				cant->spans.insert(cant->spans.end(), seg->spans.begin(), seg->spans.end());
+			} else {
+				Logger::Error("Unsupported");
+				return nullptr;
+			}
+		} else {
+			Logger::Error("Unsupported");
+			return nullptr;
+		}
+	}
+
+	auto composition = [gradient, cant](double u)->Eigen::Matrix4d {
+		auto xyz = gradient->evaluate(u);
+		auto c = cant->evaluate(u);
+      c.col(3)(0) = 0;
+      std::swap(c.col(3)(1),c.col(3)(2));
+      Eigen::Matrix4d m;
+      m = xyz * c;
+      return m;
+	};
+
+	std::array<taxonomy::piecewise_function::ptr, 2> both = { gradient , cant };
+	// @todo: rb - this constrains the range of u to the minimum of gradient and cant
```

Comment:
Maybe also a setting? I'll work on that part.

Reply:
I don't think a setting is the solution here. The same issue exists with IfcGradientCurve where the horizontal and vertical curves could differ in length. We discussed earlier that maybe u-values beyond the end of a curve would yield NAN results. In that case we want to set the span to the maximum length instead of minimum length.

The question of cases with cant and gradient curves being longer or shorter and beginning before or after the composite curve has been posted to the IF miro board. Perhaps it is best to wait for that discussion to happen before changing the current approach.

Reply:
Ok, while I agree that NaNs are maybe not so useful to the end user, but yielding 0s for those parts, so no super elevation, or no elevation just coincident with the horizontal, I imagined as potentially useful.